### DISCUSSION TITLE
Strange concat behavior?

### DISCUSSION THREAD

**Discussion by vlulla** [duckdb/duckdb#12441] [Q&A] [answered]
Consider this statement: `select 1 || 2 as tst;`. Here's the behavior i observed:

- duckdb: `12` (text)
- SQLite: `12` (text)
- PostgreSQL: `ERROR:  42883: operator does not exist: integer || integer` [^postgresql-err]
- BQ: `12` (text)

I am not surprised by SQLite's behavior. But I am very surprised by both [duckdb](https://duckdb.org/docs/sql/functions/numeric) and [BQ's](https://cloud.google.com/bigquery/docs/reference/standard-sql/operators#concatenation_operator) behavior. Duckdb does not even list `||` as a numeric operator but somehow accepts concatenation operator. I just stumbled upon this by accident and am ok with this behavior but am curious: should duckdb make an effort to disallow certain operations? So, for instance instead of `||` allowing args of any types [^concat-args] _one of the argument has to be a `STRING | LIST | ARRAY`_?

[^postgresql-err]: `undefined function` according to PostgreSQL [docs](https://www.postgresql.org/docs/current/errcodes-appendix.html).
[^concat-args]: `SELECT function_name, parameters, parameter_types from duckdb_functions() where function_name='||';`

**Comment by Tishj** [comment]
`concat` converts to string

**Reply by vlulla** [reply]
Hmm...interesting! Is this auto casting? I had asked a [similar question](https://github.com/duckdb/duckdb-web/issues/2791) on duckdb-web. Is this concat behavior related/similar to the _implicit casting_ that I mention in that question?

**Reply by Tishj** [reply]
```
D select * from duckdb_functions() where function_name = '||';
┌───────────────┬──────────────┬─────────────┬───────────────┬───────────────┬─────────────┬─────────┬───────────────────────┬─────────────┬──────────────┬─────────────────┬─────────┬──────────────────┬──────────────────┬──────────┬──────────────┬─────────┬────────────┐
│ database_name │ database_oid │ schema_name │ function_name │ function_type │ description │ comment │         tags          │ return_type │  parameters  │ parameter_types │ varargs │ macro_definition │ has_side_effects │ internal │ function_oid │ example │ stability  │
│    varchar    │   varchar    │   varchar   │    varchar    │    varchar    │   varchar   │ varchar │ map(varchar, varchar) │   varchar   │  varchar[]   │    varchar[]    │ varchar │     varchar      │     boolean      │ boolean  │    int64     │ varchar │  varchar   │
├───────────────┼──────────────┼─────────────┼───────────────┼───────────────┼─────────────┼─────────┼───────────────────────┼─────────────┼──────────────┼─────────────────┼─────────┼──────────────────┼──────────────────┼──────────┼──────────────┼─────────┼────────────┤
│ system        │ 0            │ main        │ ||            │ scalar        │             │         │ {}                    │ ANY         │ [col0, col1] │ [ANY, ANY]      │         │                  │ false            │ true     │          196 │         │ CONSISTENT │
└───────────────┴──────────────┴─────────────┴───────────────┴───────────────┴─────────────┴─────────┴───────────────────────┴─────────────┴──────────────┴─────────────────┴─────────┴──────────────────┴──────────────────┴──────────┴──────────────┴─────────┴────────────┘
```

It lists ANY as input

**Reply by vlulla** [reply]
Yes...at least, for `||` we can see that it takes `ANY` but it is unclear that implicit casting will happen for other operators like `>`,`<` which are not mentioned in the duckdb_functions table. For instance, how would i have known of the implicit casting behavior displayed in the below statement had i not stumbled upon it? The docs appear to me unclear about these edge cases.

```
D select 11<9, '11'<9,11<'9','11'<'9';
┌──────────┬────────────┬────────────┬──────────────┐
│ (11 < 9) │ ('11' < 9) │ (11 < '9') │ ('11' < '9') │
│ boolean  │  boolean   │  boolean   │   boolean    │
├──────────┼────────────┼────────────┼──────────────┤
│ false    │ false      │ false      │ true         │
└──────────┴────────────┴────────────┴──────────────┘
```

**Reply by Tishj** [reply]
I think that behavior can be attributed to <https://github.com/duckdb/duckdb/pull/10115>
Namely:
> String literals are treated as a separate type from VARCHAR - and can be implicitly cast to any other type where-ever they are used

I don't think this has the same behavior if you first create a column out of it

**Reply by Tishj** [reply, marked as answer]
```
D create table tbl as select 11 a, 9 b, '11' c, '9' d;
D select * from tbl;
┌───────┬───────┬─────────┬─────────┐
│   a   │   b   │    c    │    d    │
│ int32 │ int32 │ varchar │ varchar │
├───────┼───────┼─────────┼─────────┤
│    11 │     9 │ 11      │ 9       │
└───────┴───────┴─────────┴─────────┘
D select a<b, c<b, a<d, c<d from tbl;
Binder Error: Cannot compare values of type VARCHAR and type INTEGER - an explicit cast is required
LINE 1: select a<b, c<b, a<d, c<d from tbl;
```

**Reply by vlulla** [reply]
Very nice! Therefore, it appears that both the operands have to be columns! I think this ought to be mentioned explicitly in the documentation somewhere. Many thanks @Tishj for clarifying this for me!
